gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java

```diff
@@ -125,30 +176,23 @@ private int getThreadPoolSize() {
   }
 
   @Override
-  public void compact() throws IOException {
+  public void compact() throws Exception {
     try {
       copyDependencyJarsToHdfs();
-      Set<String> topics = findAllTopics();
+      List<String> topics = findAllTopics();
       processTopics(topics);
+      checkForThrowables();
+    } catch (Throwable t) {
 
+      // This throwable is logged here before propagated. Otherwise, if another throwable is thrown
+      // in the finally-block, this throwable will be suppressed.
+      LOG.error("Caught throwable during compaction", t);
+      throw Throwables.propagate(t);
     } finally {
-      this.executorService.shutdown();
       try {
-        executorService.awaitTermination(getMRJobTimeoutValue(), TimeUnit.MINUTES);
+        shutdownExecutors();
+      } finally {
         deleteDependencyJars();
-        for (Future<?> future : this.futures) {
```

Comment:
Instead of storing `Throwable`s in `compactionJobThrowables` and check them in `checkForThrowables`, you can probably use `ListeningFuture` and `ListenableFuture` and the callback mechanism around it. Take a look at https://github.com/google/guava/wiki/ListenableFutureExplained.